core/src/test/java/org/apache/calcite/test/RelOptRulesTest.java

```diff
@@ -4292,6 +4293,20 @@ private Sql checkSubQuery(String sql) {
     checkPlanning(program, sql);
   }
 
+  @Test public void testConvertMultiJoinRuleLoptOptimizeJoinRuleNonOptimized() throws Exception {
```

Comment:
Does this really validate that the rule was executed only once?
It does not look so. At least from the brief analysis of the test code.